### PR TITLE
first version of MCH preclustering and associated structures

### PR DESCRIPTION
**Pull request by pillot**
Hi,

The preclusterizer itself is already quite optimised, but the data structures, their serialization/deserialization, as well as the mapping are going to evolve.

Since it is the first MUON commit it also defines the directory structure, with 2 subdirectories, one for MCH and one (to come) for MID.

Cheers,
Philippe

### CODE REVIEW

**Comment by shahor02** [comment]
Hi @pillot, we are gradually moving the classes for data formats to O2/DataFormats/... folders (for particular detector classes: DataFormats/Detectors/<DetName>), keeping in the O2/Detectors only processing classes which depend on DataFormats but not vice versa. 
Should not the classes like DigitBlock or PreClusterBlock belong to DataFormats?

Cheers,
 Ruben

**Comment by pillot** [comment]
Hi Ruben,
Ultimately, yes, but the main purpose of the current classes is to interface with data format defined in HLT code. I have 2 HLT components that I run through the wrapper to send digits and read preclusters in those formats, which I use to test the preclustering. So I would say that digit and precluster data format for MCH in O2 are not yet well defined.
Cheers,
Philippe

**Comment by shahor02** [comment]
Hi Philippe,
OK, but take into account that moving the files later is quite complicated with O2 dependencies model, requires manual editing cmake/O2Dependencies.cmake and all files depending on the moved header (since we include also the directory). I was commenting only on the files location, the content may evolve w/o problem.
Cheers,
 Ruben

**Comment by pillot** [comment]
Hi Ruben,

Well, whether I do it now or later it will be the same work ;-). But I think once we will define an O2 data format we will create new files and those ones will be deprecated eventually, or kept only for test purpose, which is why I am a bit reluctant to move the files now.

Actually, only the DigitStruct defined in DigitBlock is something that we could use for O2 MCH digit format. The DigitBlock itself is only to interface with HLT format. The PreClusterBlock is a custom serializer/deserializer that write/read digits grouped in preclusters on the buffer as done in the HLT code, but we don't have a precluster object in itself (appart for internal use in the preclusterizer). I think this will also change at some point to use more standard serializer and probably the data structure in the buffer will change as well.

Cheers,
Philippe

**Comment by sawenzel** [comment]
I am partially confused. Could you clarify why you commit these files to O2 ... if they are not used for O2? What is the purpose of this PR overall?

File: Detectors/MUON/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+# * Contributors are mentioned in the code where appropriate.              *
+# *                                                                        *
+# * Permission to use, copy, modify and distribute this software and its   *
+# * documentation strictly for non-commercial purposes is hereby granted   *
+# * without fee, provided that the above copyright notice appears in all   *
+# * copies and that both the copyright notice and this permission notice   *
+# * appear in the supporting documentation. The authors make no claims     *
+# * about the suitability of this software for any purpose. It is          *
+# * provided "as is" without express or implied warranty.                  *
+# **************************************************************************
```
Comment (sawenzel) [inline Detectors/MUON/CMakeLists.txt]: I believe the copyright notice is different now.

**Comment by pillot** [comment]
The preclusterizer is for O2 and this is the main purpose of this commit. But I need digits and preclusters to make it work and test it. For the moment I kept the data structure that I developed some time ago in HLT, also to profit from the HLT components that publish those digits and analyse those preclusters. My point is simply that these data formats will change at some point so it is a bit premature to move then in the O2/DataFormats. But I need something to start with.

File: Detectors/MUON/CMakeLists.txt (hunk 1 above)
Comment (pillot) [reply Detectors/MUON/CMakeLists.txt]: Right! I copied/past it from Detectors/CMakeLists.txt and did pay attention to this. I will commit a fix.

**Pull request merged**